functions/src/stories.ts

```diff
@@ -11,83 +11,161 @@ const index = algoliasearch(
+        await index.partialUpdateObject(
+            {objectID: translationID, ...entry},
+            {createIfNotExists: true}
+        );
+      } else {
         await index.saveObject(
-        {
-          storyID,
-          storyAuthorID: story.metaData.authorId,
-          translationID,
-          translationAuthorID: translation.metaData.authorId,
-          language: translation.language,
-          title,
-          description,
-          tags,
-          likes: translation.metaData.likes,
-        } as storyRecord,
-        {autoGenerateObjectIDIfNotExist: true}
+            entry,
+            {autoGenerateObjectIDIfNotExist: true}
         );
       }
     });
 
-// TODO
+export const trackStoryUpdateTime = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}/assets/{assetID}"
+    )
+    .onWrite(async (change, context) => {
+      if (!change.after.exists) return;
+      await storyDoc(context.params.storyID, context.params.translationID)
+          .update({
+            "metaData.lastUpdateAt": firestore.FieldValue.serverTimestamp(),
+          });
+    });
+
+export const updateStoryMeta = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}"
+    )
+    .onWrite(async (change, context) => {
+      if (change.after.exists) {
+        const meta = change.after.data()?.metaData;
+        await index.partialUpdateObject({
+          objectID: context.params.translationID,
+          likes: meta.likes ?? 0,
+          views: meta.views ?? 0,
+          lastUpdateAt: meta.lastUpdateAt,
+          imageUrl: meta.imageUrl,
+        });
+      }
+    });
+
+export const indexTrendingStories = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 72 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: ["storyID"],
+        batch: async (batch) => {
+          for (const hit of batch) {
+            const {storyID, objectID} = hit;
+            const doc = storyDoc(storyID, objectID);
+            const meta = await doc.get().then((r) => r.data()!.metaData);
+            const trending = (meta.views ?? 0) - (meta.lastIndexedViews ?? 0);
+            await doc.update({"metaData.lastIndexedViews": meta.views ?? 0});
+            await index.partialUpdateObject({objectID, trending});
+          }
+        },
+      });
+    });
+
+export const indexRandomExploration = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 24 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: [],
+        batch: async (batch) => {
+          for (const hit of batch) {
+            await index.partialUpdateObject({
+              objectID: hit.objectID,
+              explore: Math.random(),
```

Comment:
Is this field supposed to say whether a story should go under the "explore" section? Why is it random and not depending on the story creation date?

Reply:
This is a rotating value to bring some random (but still sorted by likes & views) to the front page.

Reply:
Then I guess the UI shouldn't say "Explore **new**"

Reply:
The intention was to give users some random new stories they might have missed otherwise.

Reply:
Understood, but still feels like "New" is not entirely accurate

Reply:
I've meant "new" not by update date but by past experience, assuming that this randomization will bring stories the user hasn't played before. But I see that it might be ambiguous, so it'll be just "explore".